contracts/SDUtilityPool.sol

```diff
@@ -0,0 +1,906 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+pragma solidity 0.8.16;
+
+import './library/UtilLib.sol';
+import './interfaces/IStaderConfig.sol';
+import './interfaces/IStaderOracle.sol';
+import './interfaces/ISDIncentiveController.sol';
+import './interfaces/ISDUtilityPool.sol';
+import './interfaces/SDCollateral/ISDCollateral.sol';
+import './interfaces/IPoolUtils.sol';
+import './interfaces/IOperatorRewardsCollector.sol';
+
+import '@openzeppelin/contracts/utils/math/Math.sol';
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol';
+import '@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol';
+
+contract SDUtilityPool is ISDUtilityPool, AccessControlUpgradeable, PausableUpgradeable {
+    using Math for uint256;
+
+    uint256 public constant DECIMAL = 1e18;
+
+    uint256 public constant MAX_PROTOCOL_FEE = 1e17; // 10%
+
+    // State variables
+
+    /// @notice Fraction of fee currently set aside for protocol
+    uint256 public protocolFeeFactor;
+
+    /// @notice Block number that fee was last accrued at
+    uint256 public accrualBlockNumber;
+
+    /// @notice Accumulator of the total earned interest rate since start of pool
+    uint256 public utilizeIndex;
+
+    /// @notice Total amount of outstanding SD utilized
+    uint256 public totalUtilizedSD;
+
+    /// @notice Total amount of protocol fee
+    uint256 public accumulatedProtocolFee;
+
+    /// @notice utilization rate per block
+    uint256 public utilizationRatePerBlock;
+
+    /// @notice value of cToken supply
+    uint256 public cTokenTotalSupply;
+
+    /// @notice upper cap on ETH worth of SD utilized per validator
+    uint256 public maxETHWorthOfSDPerValidator;
+
+    /// @notice request ID to be finalized next
+    uint256 public nextRequestIdToFinalize;
+
+    /// @notice request ID to be assigned to a next withdraw request
+    uint256 public nextRequestId;
+
+    /// @notice amount of SD requested for withdraw
+    uint256 public sdRequestedForWithdraw;
+
+    /// @notice batch limit on withdraw requests to be finalized in single txn
+    uint256 public finalizationBatchLimit;
+
+    /// @notice amount of SD reserved for claim request
+    uint256 public sdReservedForClaim;
+
+    /// @notice minimum block delay between requesting for withdraw and claiming
+    uint256 public undelegationPeriodInBlocks;
```

Comment:
Why did we need an undelegation period separate from finalization timedelay?

Reply:
lets chat about this